clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -2025,29 +2046,43 @@ scheduler_tmpl_t<RecordType, ReaderType>::next_record(output_ordinal_t output,
                         input->switch_to_input = it->second;
                     }
                 } else if (record_type_is_instr(record)) {
-                    // Assume it will block and we should switch to a different input.
-                    need_new_input = true;
-                    in_wait_state = true;
+                    if (treat_syscall_as_blocking(input)) {
+                        // Model as blocking and should switch to a different input.
+                        need_new_input = true;
+                        in_wait_state = true;
+                        VPRINT(this, 3,
+                               "next_record[%d]: hit blocking syscall in input %d\n",
+                               output, input->index);
+                    }
+                    input->processing_syscall = false;
                     input->processing_blocking_syscall = false;
-                    VPRINT(this, 3, "next_record[%d]: hit blocking syscall in input %d\n",
-                           output, input->index);
+                    input->pre_syscall_timestamp = 0;
                 }
-            } else if (record_type_is_marker(record, marker_type, marker_value) &&
-                       marker_type == TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL) {
+            }
+            if (record_type_is_marker(record, marker_type, marker_value) &&
+                marker_type == TRACE_MARKER_TYPE_SYSCALL) {
+                input->processing_syscall = true;
+                input->pre_syscall_timestamp = input->reader->get_last_timestamp();
+            }
+            if (record_type_is_marker(record, marker_type, marker_value) &&
+                marker_type == TRACE_MARKER_TYPE_MAYBE_BLOCKING_SYSCALL) {
                 input->processing_blocking_syscall = true;
-            } else if (options_.quantum_unit == QUANTUM_INSTRUCTIONS &&
-                       record_type_is_instr(record)) {
+                // Generally we should already have the timestamp from a just-prior
+                // syscall marker, but we support tests and other synthetic sequences
+                // with just a maybe-blocking.
+                input->pre_syscall_timestamp = input->reader->get_last_timestamp();
+            }
+            if (options_.quantum_unit == QUANTUM_INSTRUCTIONS &&
+                record_type_is_instr(record)) {
                 ++input->instrs_in_quantum;
```

Comment:
So before we were not incrementing instrs_in_quantum when processing_blocking_syscall and record_type_is_instr, but we should have? And now we are because we s/`else if`/`if`?

Reply:
It does appear that way.  It doesn't seem to have affected any tests, even the ones with tiny quanta.  The elses caused other problems too with the changes in this PR so they had to go.